### PR TITLE
[Feat] Add GQA Sink Backward Varlen

### PR DESCRIPTION
**Pull request by wch0810**
## Description

Refactor GQA + Attention Sink Flash Attention Backward (Varlen) into a **single-kernel backward**. All backward computation (Delta + 5-GEMM + 2-softmax-bwd + Compensated GEMM + dSinks) is now fused into one `flashattn_bwd_single` kernel launch.

Strategy: single kernel with `T.serial` K-loop, on-chip direct CV handoff (L0C↔UB, independent buffer per handoff), Compensated GEMM via L0C `init=False` accumulation (no separate correction kernel), dSinks computed in-kernel (Phase 6). Uses `combineCV` + `auto_sync` passes (no manual `T.Scope`/`T.barrier_all`/`cross_flag`).

Tested on Ascend NPU, fp16, B=8, H=64, G=16, q=k=2048, D=128, causal. Precision verified against PyTorch golden using 169-line mixed-tolerance standard (atol=6.10e-5, rtol=1.95e-3, ratio=0.99).

## Performance (B=8, H=64, G=16, q=k=2048, D=128, fp16, causal)
| Kernel                              | Q seq | backward launches | Latency (ms) | TFLOPS | vs GPU  | Max Diff |
| ----------------------------------- | ----- | ----------------- | ------------ | ------ | ------- | -------- |
|  Forward                    | 2048  | 1                 | 93.35        | 5.89   | —       | 1.95e-03 |
|  Backward (single kernel)   | 2048  | **1**             | **114.0**    | 12.06  | 4.0x    | 1.95e-03 |
|  Fwd+Bwd (e2e)              | 2048  | 2                 | 207.3        | 9.28   | —       | —        |
| GPU baseline (backward only)        | 2048  | 4                 | 28.574       | —      | 1.0x    | —        |

- Previous 5-kernel version: 34 launches, 160.9ms, 5.6x. Single-kernel is **29% faster**.
- All intermediates on-chip: **0 MB GM workspace** (was ~512MB).

## Precision (169-line standard, all 4 outputs blocking)
| Output  | matched_ratio | max_abs   | dtype | Result |
| ------- | ------------- | --------- | ----- | ------ |
| O       | 0.998         | 9.77e-04  | fp16  | PASS   |
| Delta   | 1.000         | 1.31e-06  | fp32  | PASS   |
| dQ      | 0.999         | 9.77e-04  | fp16  | PASS   |
| dK      | 0.999         | 1.95e-03  | fp16  | PASS   |
| dV      | 1.000         | 9.77e-04  | fp16  | PASS   |
| dSinks  | 1.000         | 2.38e-07  | fp32  | PASS   |

- L0: 8/8 PASS, determinism 7/7 bit-exact
- L1: 13/13 PASS, L2: 10/10 PASS
image

## Notes
- **Single-kernel backward**: K-loop `T.serial` inside kernel, no host for-loop, no `torch.npu.synchronize()` between stages.
- **p_delta/ds_delta fp16 UB→L1 direct** (8KB each): eliminates GM roundtrip for Compensated GEMM residuals.
- **P_fp32 retention**: `s_ub` retains P_fp32 across phases (Phase 2 → Phase 4), eliminating GM roundtrip for dS compute.
- **dSinks in-kernel**: Phase 6 `T.tile.exp`, fp32 output. Golden recomputes using kernel's lse/Delta to isolate `T.tile.exp` precision (ratio=1.0).
- **dQ/dK/dV fp32 atomic_add + host `.half()`**: GQA groups=16 requires fp32 cross-head accumulation (fp16 atomic_add precision insufficient).
- Block sizes: BLOCK_M=64, BLOCK_N=64, threads=1.
- Techniques adapted from `example_gqa_sink_bwd_bhsd` (rev3 single-kernel pattern).
- Limitations: `threads=1` (Ascend `AUTO_CV_SYNC` does not support `threads=2` for hybrid kernels with `atomic_add` — workspace race). This is the primary source of the 4.0x gap vs GPU (128 threads). The single-kernel on-chip architecture is correct; performance will improve once the framework supports multi-thread CV fusion.

### CODE REVIEW

File: examples_experiment/example_gqa_sink_bwd_varlen/test_gqa_sink_bwd_varlen.py [hunk 1]
```diff
@@ -0,0 +1,958 @@
...
+        qp, qr, qm = check_precision(dQ_fp16[..., :D].cpu(), dQ_ref.cpu(), DTYPE_FP16)
+        min_ratio = min(min_ratio, qr)
+        max_diff = max(max_diff, qm)
+        all_passed &= qp
+        if not qp:
+            raise AssertionError(f"dQ precision failed: ratio={qr:.4f}, max_abs={qm:.3e}")
+        for b in range(B):
+            ks = cu_seqlens_k[b]
+            ke = cu_seqlens_k[b + 1]
+            # dK/dV 从 host .half() 来 (fp32 atomic_add + host cast)
```
Comment (hedi515) [inline examples_experiment/example_gqa_sink_bwd_varlen/test_gqa_sink_bwd_varlen.py]: 英文注释

File: examples_experiment/example_gqa_sink_bwd_varlen/example_gqa_sink_bwd_varlen.py [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+        # Verify outputs are non-trivial (non-zero, finite)
+        print(f"  O      : max_abs={O.abs().max().item():.3e}")
+        print(f"  dQ     : max_abs={dQ.abs().max().item():.3e}")
+        print(f"  dK     : max_abs={dK.abs().max().item():.3e}")
+        print(f"  dV     : max_abs={dV.abs().max().item():.3e}")
+        print(f"  dSinks : max_abs={dSinks.abs().max().item():.3e}")
+
+        ok = all(torch.isfinite(t).all().item() and t.abs().max().item() > 0 for t in [O, dQ, dK, dV, dSinks])
+
+        print("\nTest Passed!" if ok else "\nTest FAILED")
```
Comment (hedi515) [inline examples_experiment/example_gqa_sink_bwd_varlen/example_gqa_sink_bwd_varlen.py]: 格式改一下，统一形式

**Comment by hedi515** [comment]
lgtm

File: examples_experiment/example_gqa_sink_bwd_varlen/example_gqa_sink_bwd_varlen.py (hunk 2 above)
Comment (wch0810) [reply examples_experiment/example_gqa_sink_bwd_varlen/example_gqa_sink_bwd_varlen.py]: Format has been changed for consistent styling.

File: examples_experiment/example_gqa_sink_bwd_varlen/test_gqa_sink_bwd_varlen.py (hunk 1 above)
Comment (wch0810) [reply examples_experiment/example_gqa_sink_bwd_varlen/test_gqa_sink_bwd_varlen.py]: All comments have been updated to English.

**Review by ChaoyangJi — APPROVED** [review]
approve

**Review by LLMZhangYC — APPROVED** [review]
approve
